[Categories] Display product categories on Product Edit screen

Ecarrion opened the pull request:
Relates to: #2000 

# Why

- This PR displays(under the `editProductsRelease3`)  the product categories in the product edit screen.
- This PR also fixes some icon orientation issues on RTL languages.

# How

- Adds a `.categories` case in the `ProductFormSection.SettingsRow` enum.
- Adds a `categoriesDescription` helper method on `Product` to display a comma separated list of category names.
- Added the necessary code to include the new categories section on `DefaultProductFormTableViewModel`
- Added tests

# Test steps
- Go to the Products tab
- Tap on a simple Product with no categories
   - See categories row with "Uncategorized" description label
- Tap con a simple Product with one category
   - See categories row with the single category in the description label
- Tap con a simple Product with multiple categories
   - See categories row with the formatted list of categories names in the description label

# Notes
 - ~I have not added a categories list placeholder because the web app doesn't let you have a 0 category product. It always have at least the `uncategorized` category. Let me know if this is wrong~ 
I have added a categories list placeholder("Uncategorized") for when a product doesn't have any category.

- I have copied the approach with `DefaultProductFormTableViewModel_EditProductsM2Tests`. This seems a bit hard to maintain but I hope that `EditProductsM2Tests` should be shipped soon?

# Screenshots

Single Category|Mutiple Categories|RTL
---|---|---
<img src="https://user-images.githubusercontent.com/562080/77609229-6dffb300-6eed-11ea-8670-5e93451c8459.png" width="300" />|<img src="https://user-images.githubusercontent.com/562080/77609234-6fc97680-6eed-11ea-9fc1-5d2020e6b59d.png" width="300" />|<img src="https://user-images.githubusercontent.com/562080/77609232-6f30e000-6eed-11ea-9684-0aca60259378.png" width="300" />

Ecarrion commented:
@shiki Could you help me setting labels, reviewers and milestones please? 🙏 
Also, let me know if I need to do some extra process in order to get those new texts translated.

----

Tagging @jaclync too, since you worked in this screen recently maybe you want to chime in in the review 🙂

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product+ProductForm.swift:
```diff
@@ -16,4 +16,15 @@ extension Product {
         return description.removedHTMLTags.trimmingCharacters(in: .whitespacesAndNewlines)
     }
+
+    /// Returns a coma separated string with each category names.
```
nit: comma separated :)

jaclync commented on WooCommerce/Classes/Extensions/UIImage+Woo.swift:
```diff
@@ -47,6 +47,12 @@ extension UIImage {
 
+    /// Product categories Icon
+    ///
+    static var categoriesIcon: UIImage {
```
when a new icon helper is created, we usually add a test case in [`IconsTests`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/WooCommerceTests/Extensions/IconsTests.swift) - could we also add one for this new icon?

Ecarrion replied:
fixed https://github.com/woocommerce/woocommerce-ios/pull/2043/commits/b1983f1b670fa66d63adcc71eb25bea4cd3ee3bc

jaclync commented on WooCommerce/Classes/Extensions/UIImage+Woo.swift:
```diff
@@ -36,7 +36,7 @@ extension UIImage {
     ///
     static var briefDescriptionImage: UIImage {
-        return UIImage.gridicon(.alignLeft, size: CGSize(width: 24, height: 24))
+        return UIImage.gridicon(.alignLeft, size: CGSize(width: 24, height: 24)).imageFlippedForRightToLeftLayoutDirection()
```
this is very thoughtful!

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/DefaultProductFormTableViewModel+EditProductsM3Tests.swift:
```diff
@@ -0,0 +1,173 @@
+
+    private let mockFeatureFlagService = MockFeatureFlagService(isEditProductsRelease2On: true, isEditProductsRelease3On: true)
+
+    func testViewModelForSimplePhysicalProductWithoutImagesWhenM3FeatureFlagIsOff() {
```
sorry I think this is from a mistake in the [M2 tests](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/WooCommerceTests/ViewRelated/Products/Edit%20Product/DefaultProductFormTableViewModel+EditProductsM2Tests.swift#L13), where the M3 feature flag isn't off as described in the test function name (`*WhenM3FeatureFlagIsOff`). I think `WhenM3FeatureFlagIsOff` can just be removed from this function name?

Ecarrion replied:
fixed https://github.com/woocommerce/woocommerce-ios/pull/2043/commits/67d73521ae6e6696ac8fa809f25d8d2d75e3a3a8

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product+ProductForm.swift:
```diff
@@ -16,4 +16,15 @@ extension Product {
+    var categoriesDescription: String {
+        let categoriesNames = categories.map { $0.name }
+        if #available(iOS 13.0, *) {
+            return ListFormatter.localizedString(byJoining: categoriesNames)
```
cool usage of `ListFormatter` (TIL!)

Ecarrion commented:
> I can have products without any categories and the row just shows the placeholder text ("Categories") as in the screenshots below

@jackiejade Thank you for double checking that, I have added a proper placeholder. 
I've also added the **Test steps** section

🙇

Garance91540 commented:
Hi! Looking great, just a few questions:

When are we planning of adding the "more details" button? @jaclync @shiki 
The category section should only appears once you've selected it in the "More details" section: https://d.pr/i/oTwcA2. I wanted to make sure this is how we were thinking of implementing it!

When a user has selected the category section in the More details menu and there's no category, could you share with me screenshot of the selection screen + how it looks like when you land back on the product screen?

Ecarrion commented:
> When a user has selected the category section in the More details menu and there's no category, could you share with me screenshot 

This is how it looks when you select a product without categories

<img width="383" alt="uncategorized" src="https://user-images.githubusercontent.com/562080/77658679-070de880-6f45-11ea-8163-d81806645731.png">

jaclync commented:
> When are we planning of adding the "more details" button? @jaclync @shiki
The category section should only appears once you've selected it in the "More details" section: https://d.pr/i/oTwcA2. I wanted to make sure this is how we were thinking of implementing it!

I sort of remember Kylea mentioned that we show all the fields when editing a product, and show minimal fields when creating a product? in any case, the design has the "+ Add more details" CTA at the bottom for both adding & editing a product.

The major reason we haven't implemented the "+ Add more details" CTA is because we don't have a bottom sheet framework yet, and the native action sheet wouldn't look good with the design with a title and subtitle. To add a bottom sheet framework, we probably will go with a third-party option and this requires team discussion.

<img width="226" alt="Screen Shot 2020-03-27 at 11 50 01 AM" src="https://user-images.githubusercontent.com/1945542/77720051-c3d57780-7021-11ea-840d-e37f505233fc.png">

I created 2 tickets for M2, @Garance91540 lemme know if you think M3 or another milestone might be more appropriate!
- [Add a bottom sheet framework (TBD)](https://github.com/woocommerce/woocommerce-ios/issues/2054)
- [Add the more details CTA that brings up the bottom sheet options](https://github.com/woocommerce/woocommerce-ios/issues/2053)

Garance91540 commented:
>I sort of remember Kylea mentioned that we show all the fields when editing a product, and show minimal fields when creating a product? in any case, the design has the "+ Add more details" CTA at the bottom for both adding & editing a product.

Ah thanks for the context @jaclync  I don't think I was aware of this. If we can implement the "more details" for edit + creation mode that'd be my preference, just to reduce cognitive load on the product page (which [can get quite busy](https://d.pr/i/vhQI54)). But it's not a top priority so implementing it when we have the bottom sheet component ready is fine!

shiki commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Product+ProductFormTests.swift:
```diff
@@ -18,11 +18,37 @@ class Product_ProductFormTests: XCTestCase {
+        let product = sampleProduct(categories: categories)
+        let expectedDescription: String = {
+            if #available(iOS 13.0, *) {
+                return ListFormatter.localizedString(byJoining: ["Pants", "Dress", "Shoes"])
```
I'm still not sure how to feel about tests like this. We're simply just repeating the implementation of `.categoriesDescription`. It would probably have been nicer if `ListFormatter.localizedString()` allows us to pass a `locale` so then we could use a full string for comparison instead of having this call here. But it doesn't. ¯\\\_(ツ)\_/¯

Ecarrion replied:
I wasn't happy with this test either and I turns out that we can pass a locale to format the string.
I've updated the code here https://github.com/woocommerce/woocommerce-ios/pull/2043/commits/44e0467c64ef2365ba658b6262b658ad5cf04cb4

shiki commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/DefaultProductFormTableViewModel.swift:
```diff
@@ -177,6 +181,13 @@ private extension DefaultProductFormTableViewModel {
+    func categoriesRow(product: Product) -> ProductFormSection.SettingsRow.ViewModel {
+        let icon = UIImage.categoriesIcon
+        let title = Constants.categoriesTitle
+        let details = product.categoriesDescription.isNotEmpty ? product.categoriesDescription: Constants.categoriesPlaceholder
```
You can also do:

```suggestion
        let details = product.categoriesDescription ?? Constants.categoriesPlaceholder
```

shiki commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/DefaultProductFormTableViewModel+EditProductsM3Tests.swift:
```diff
@@ -0,0 +1,173 @@
+/// The same tests as `DefaultProductFormTableViewModel_EditProductsM2Tests`, but with Edit Products M2 and M3 feature flag on.
+/// When we fully launch Edit Products M2 and M3, we can replace `DefaultProductFormTableViewModel_EditProductsM2Tests` with the test cases here.
+///
+final class DefaultProductFormTableViewModel_EditProductsM3Tests: XCTestCase {
```
To answer this question:

> I have copied the approach with DefaultProductFormTableViewModel_EditProductsM2Tests. This seems a bit hard to maintain but I hope that EditProductsM2Tests should be shipped soon?

I think the tests in this class are testing too many things and are not specific enough. Perhaps what we could have done here for the "category" is only test things like 

```swift
fund testViewModelWillShowCategoriesWhenM3FeatureFlagIsOn()
```

And only test this portion:

```swift
            if case .categories(_) = rows[3] {} else {
                XCTFail("Unexpected setting section: \(rows[3])")
            }
```

But no matter, let's leave this as is. 🙂

Ecarrion commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product+ProductForm.swift:
```diff
@@ -16,4 +16,15 @@ extension Product {
+        if #available(iOS 13.0, *) {
+            return ListFormatter.localizedString(byJoining: categoriesNames)
+        } else {
+            return categoriesNames.joined(separator: ",")
```
🤦‍♂ 
Fixed https://github.com/woocommerce/woocommerce-ios/pull/2043/commits/10396a8e84d05a0d7c03190f9396c82f4f0e6737

Ecarrion commented:
@shiki I've fixed the requested change and also improved a bit the `categoriesDescription` method so it's more testable. Please ♻️

jaclync commented:
> If we can implement the "more details" for edit + creation mode that'd be my preference, just to reduce cognitive load on the product page (which can get quite busy). But it's not a top priority so implementing it when we have the bottom sheet component ready is fine. I'm also seeing bottom sheets everywhere on the WP app, is there a way we could use theirs?

@Garance91540 yes it'd be nice to hide the optional rows on the product page for both editing and creating a product. Could you share a screenshot of the bottom sheet in the WP app? I checked the latest WP app and didn't find any action that brings out a custom bottom sheet (just the native action sheet), but maybe I missed the actions that trigger it 😄

Garance91540 commented:
> Could you share a screenshot of the bottom sheet in the WP app? 

Just sent a bunch in DM!
